lib/eco/overlay.py

```diff
@@ -76,10 +76,30 @@ def add_railroad_datum(self, datum):
         for key in datum:
             self._railroad_data[key] = datum[key]
 
+    def add_railroad_data(self, data):
+        for datum in data:
+            for key in datum:
+                self._railroad_data[key] = datum[key]
+
     def clear_data(self):
         self._heatmap_data = dict()
         self._railroad_data = dict()
 
+    def _get_visible_lines(self, viewport_height):
+        """Return the line numbers of the first and last visible lines of code.
+        """
+        start_line = self.node_editor.paint_start[0] + 1
+        end_line = start_line
+        total_height = 0.0
+        gfont = QApplication.instance().gfont  # Current system font.
+        for line_no, line in enumerate(self.tm.lines):
+            if (line_no + 1) > start_line:
+                end_line += 1
+                total_height += line.height * (gfont.fontht)
+            if total_height > viewport_height:
+                break
+        return (start_line, end_line)
+
```

Comment:
This looks correct. I'm not 100% sure but you might be able to get `end_line` from `self.paint_nodes` which returns the last line drawn. Then you won't have to compute this number again. This of course requires that the overlay is drawn *after* the source code has been painted.